### PR TITLE
Prevent infinite recursion within Tensor.__repr__

### PR DESCRIPTION
**Pull request by aorenste**
`Tensor.__repr__` calls functions which can perform logging which ends up logging `self` (with `__repr__`) causing an infinite loop. Instead of logging all the args in FakeTensor.dispatch log the actual parameters (and use `id` to log the tensor itself).

The change to torch/testing/_internal/common_utils.py came up during testing - in some ways of running the test parts was `('test', 'test_testing.py')` and so `i` was 0 and we were doing a join on `()` which was causing an error.

Repro:
```
import torch
from torch.testing import make_tensor
from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
t = torch.sparse_coo_tensor(((0, 1), (1, 0)), (1, 2), size=(2, 2))
t2 = FakeTensor.from_tensor(t, FakeTensorMode())
print(repr(t2))
```
and run with `TORCH_LOGS=+all`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120206

### CODE REVIEW

File: torch/_tensor.py [hunk 1]
```diff
@@ -455,12 +455,19 @@ def __setstate__(self, state):
         self.requires_grad, _, self._backward_hooks = state
 
     def __repr__(self, *, tensor_contents=None):
-        if has_torch_function_unary(self):
-            return handle_torch_function(
-                Tensor.__repr__, (self,), self, tensor_contents=tensor_contents
-            )
-        # All strings are unicode in Python 3.
-        return torch._tensor_str._str(self, tensor_contents=tensor_contents)
+        _in_recursive_repr = getattr(self, "_in_recursive_repr", 0)
```
Comment (oulgen) [inline torch/_tensor.py:458]: why do you need to count here, seems like a boolean would have been enough?

File: torch/_tensor.py [hunk 2]
```diff
@@ -455,12 +455,19 @@ def __setstate__(self, state):
...
+        self._in_recursive_repr = _in_recursive_repr + 1
+        try:
+            if has_torch_function_unary(self):
+                return handle_torch_function(
+                    Tensor.__repr__, (self,), self, tensor_contents=tensor_contents
+                )
+            # All strings are unicode in Python 3.
+            return torch._tensor_str._str(self, tensor_contents=tensor_contents)
+        finally:
+            self._in_recursive_repr = _in_recursive_repr
```
Comment (oulgen) [inline torch/_tensor.py:470]: not sure i follow why you need the try finally? if any of this throws an exception, then the repr results wont be important anyway?

File: torch/_tensor.py [hunk 3]
```diff
@@ -455,12 +455,19 @@ def __setstate__(self, state):
...
+            return f"<recursive tensor {id(self)}>"
+        self._in_recursive_repr = _in_recursive_repr + 1
+        try:
+            if has_torch_function_unary(self):
+                return handle_torch_function(
+                    Tensor.__repr__, (self,), self, tensor_contents=tensor_contents
+                )
+            # All strings are unicode in Python 3.
+            return torch._tensor_str._str(self, tensor_contents=tensor_contents)
+        finally:
```
Comment (oulgen) [inline torch/_tensor.py:469]: shouldn't you unset this attr after the result is determined? otherwise the user can observe it

**Review by yanboliang — APPROVED** [review]

**Review by pearu — CHANGES_REQUESTED** [review]
This PR does not fix the underlying reason for the infinite recursion within `repr` and it introduces a "recursive tensor" concept that meaning does not apply to the tensors used in the reproducer of the issue.

Considering the traceback of the infinite recursion exception, I think the cause of the underlying problem lies in
https://github.com/pytorch/pytorch/blob/834c7a1d3ea07878ad87d127ee28606fc140b552/torch/_subclasses/fake_tensor.py#L1203
that calls `repr` within `repr` call.

Here's another way to stop the infinite recursion:
```
diff --git a/torch/_subclasses/fake_tensor.py b/torch/_subclasses/fake_tensor.py
index c3f19a94e50..995e6c2dfff 100644
--- a/torch/_subclasses/fake_tensor.py
+++ b/torch/_subclasses/fake_tensor.py
@@ -1200,8 +1200,15 @@ class FakeTensorMode(TorchDispatchMode):
 
     def dispatch(self, func, types, args=(), kwargs=None):
         kwargs = kwargs or {}
-        log.debug("%s %s %s", func, args, kwargs)
 
+        # Some attribute queries that can be serviced directly
+        # See Note [is_coalesced is dispatched]
+        if func in _DISPATCH_HANDLE_DIRECTLY:
+            # NB: no_dispatch is ok here too, this func is very simple
+            with in_kernel_invocation_manager(self):
+                return func(*args, **kwargs)
+
+        log.debug("%s %s %s", func, args, kwargs)
         if func in _DISPATCH_META_HANDLERS:
             return _DISPATCH_META_HANDLERS[func](args)
 
@@ -1787,6 +1794,8 @@ _DISPATCH_HANDLE_DIRECTLY = ordered_set(
     torch.ops.aten.is_coalesced.default,
     torch.ops.aten.dense_dim.default,
     torch.ops.aten.sparse_dim.default,
+    torch.ops.aten._indices.default,
+    torch.ops.aten._values.default,
 )
```

However, ideally, there should be a way to break the recursive dispatch calls that is initiated by the call path `repr(coo)->coo._indices()->...->log.debug(..., coo)->repr(coo)->...`. Can you find it?

**Comment by aorenste** [comment]
Hey - thanks for the feedback!

> This PR does not fix the underlying reason for the infinite recursion within `repr` and it introduces a "recursive tensor" concept that meaning does not apply to the tensors used in the reproducer of the issue.

I was purposely not trying to fix the underlying reason but rather trying a "bigger hammer" to make sure that other code that has the same problem (logging a tensor from within that tensor's `__repr__`) doesn't infinitely recurse either.  But I can see that that's not necessarily a good solution.

> Considering the traceback of the infinite recursion exception, I think the cause of the underlying problem lies in

Correct - I alluded to this in the summary where I said that it was recursive logging that was the problem.

> Here's another way to stop the infinite recursion:

This makes sense.

> However, ideally, there should be a way to break the recursive dispatch calls that is initiated by the call path `repr(coo)->coo._indices()->...->log.debug(..., coo)->repr(coo)->...`. Can you find it?

I can't tell if you know another way and are challenging me to find it or if you're just not sure and are asking. I'm pretty new to the codebase so I'm pretty unfamiliar with many of the options the code can take.  As far as I can tell the ways to break the recursion would be:
1. detect the recursion and return early (which is what this diff originally does)
2. dispatch the indices call before the log (what you suggested above)
3. do something else that causes us to not call the dispatcher - either calling aten._indices() directly or something that defeats the override? But I'm not sure what that would be.
4. Log `args[1:]` instead of `args` assuming that it's going to be `self` (or possible do an `is` check for that possibility)

**Review by pearu — CHANGES_REQUESTED** [review]
> I can't tell if you know another way and are challenging me to find it or if you're just not sure and are asking.

It is a combination of all the above because I vaguely remembered that there is a way to do that:) Here is the answer:
```
diff --git a/torch/_subclasses/fake_tensor.py b/torch/_subclasses/fake_tensor.py
index 9d2b38f40fe..6a9a4ea27f3 100644
--- a/torch/_subclasses/fake_tensor.py
+++ b/torch/_subclasses/fake_tensor.py
@@ -1184,7 +1184,8 @@ class FakeTensorMode(TorchDispatchMode):
 
     def dispatch(self, func, types, args=(), kwargs=None):
         kwargs = kwargs or {}
-        log.debug("%s %s %s", func, args, kwargs)
+        with no_dispatch():
+            log.debug("%s %s %s", func, args, kwargs)
 
         if func in _DISPATCH_META_HANDLERS:
             return _DISPATCH_META_HANDLERS[func](args)
```
and there is not need to add `_indices`/`_values` to `_DISPATCH_HANDLE_DIRECTLY` set.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @aorenste!

**Pull request closed without merging**